### PR TITLE
Add a rust pathfinding

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Replace DM node pathfinding by a rust one.
rust pr is here : https://github.com/tgstation/rust-g/pull/113

## Why It's Good For The Game

Approximately 32 times faster

## Changelog
:cl:
code: Improve performance of pathfinder
/:cl:

### CODE REVIEW

File: code/modules/ai/ai_node.dm [hunk 1]
```diff
@@ -23,12 +27,25 @@
 
 /obj/effect/ai_node/Initialize()
 	..()
-	GLOB.allnodes += src
+	GLOB.all_nodes += src
+	unique_id = id_counter++;
```
Comment (TiviPlus) [inline code/modules/ai/ai_node.dm:31]: ;

File: code/modules/ai/ai_node.dm [hunk 2]
```diff
@@ -38,7 +55,8 @@
 	weights[identifier][name] = amount
 
 /obj/effect/ai_node/Destroy()
-	GLOB.allnodes -= src
+	GLOB.all_nodes[unique_id + 1] = null
```
Comment (TiviPlus) [inline code/modules/ai/ai_node.dm:58]: Would it not be better to start the Id at 1 or is this a rust consideration?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/ai/ai_node.dm (hunk 2 above)
Comment (BraveMole) [reply code/modules/ai/ai_node.dm:58]: rust consideration

File: code/modules/ai/ai_node.dm (hunk 1 above)
Comment (BraveMole) [inline code/modules/ai/ai_node.dm:31]:
```suggestion
	unique_id = id_counter++
```

**Comment by Lumipharon** [comment]
Does this mean having a gorillion minions will actually function?

File: .gitignore [hunk 3]
```diff
@@ -28,10 +28,6 @@ tools/MapDaemon/logs/*.txt
 
 tools/BalanceTool/raw.json
 
-
-# Ignore compiled linux libs in the root folder, e.g. librust_g.so
-/*.so
```
Comment (ZeWaka) [inline .gitignore:33]: Why remove this?

**Comment by Zaffinomlb** [comment]
Tested today with 1silo 2 spawner set up. Minions moved sluggishly and there would be moments of 3-6 second pause where the minions would not react at all.

**Comment by henriquekill9576** [comment]
This makes minions pretty much useless since now they're 3x as slow as they were before

File: .gitignore (hunk 3 above)
Comment (BraveMole) [reply .gitignore:33]: Todo: unremove that

File: code/controllers/subsystem/advanced_pathfinding.dm [hunk 4]
```diff
@@ -2,12 +2,50 @@ SUBSYSTEM_DEF(advanced_pathfinding)
...
 	var/list/datum/ai_behavior/tile_pathfinding_to_do = list()
 	///List of ai_behaviour datum asking for a tile pathfinding
 	var/list/datum/ai_behavior/node_pathfinding_to_do = list()
 
+/datum/controller/subsystem/advanced_pathfinding/Initialize(start_timeofday)
+	. = ..()
+	var/list/nodes = list()
+	for(var/obj/effect/ai_node/ai_node AS in GLOB.all_nodes)
+		nodes += list(ai_node.serialize())
+	message_admins("Registering nodes : [rustg_register_nodes_astar(json_encode(nodes))]")
```
Comment (BraveMole) [inline code/controllers/subsystem/advanced_pathfinding.dm]: todo: remove that

File: tools/ci/install_rust_g.sh [hunk 5]
```diff
@@ -3,7 +3,18 @@ set -euo pipefail
...
+git checkout pathfinder
+sudo dpkg --add-architecture i386
+sudo apt update
+sudo apt-get install g++-multilib zlib1g-dev:i386 libssl-dev:i386
+rustup target add i686-unknown-linux-gnu
+export PKG_CONFIG_ALLOW_CROSS=1
+cargo build --release --target i686-unknown-linux-gnu --features pathfinder
 mkdir -p ~/.byond/bin
-wget -O ~/.byond/bin/librust_g.so "https://github.com/tgstation/rust-g/releases/download/$RUST_G_VERSION/librust_g.so"
+cp target/i686-unknown-linux-gnu/release/librust_g.so ~/.byond/bin
```
Comment (BraveMole) [inline tools/ci/install_rust_g.sh:18]: to fix that

**Comment by Zaffinomlb** [comment]
Currently TMed today, and the before mentioned issue seems solved. However the issue now seems to be that the minions do not want to move to their AI goal.

File: .github/workflows/ci_suite.yml [hunk 6]
```diff
@@ -90,9 +90,6 @@ jobs:
           mysql -u root -proot -e 'CREATE DATABASE tg_ci_prefixed;'
       - name: Install rust-g
         run: |
-          sudo dpkg --add-architecture i386
-          sudo apt update || true
-          sudo apt install -o APT::Immediate-Configure=false libssl1.1:i386
           bash tools/ci/install_rust_g.sh
```
Comment (TiviPlus) [inline .github/workflows/ci_suite.yml:93]: Do you need to readd?

File: .gitignore (hunk 3 above)
Comment (TiviPlus) [reply .gitignore:33]: Todo

File: code/controllers/subsystem/advanced_pathfinding.dm [hunk 7]
```diff
@@ -2,12 +2,50 @@ SUBSYSTEM_DEF(advanced_pathfinding)
...
+	var/list/nodes = list()
+	for(var/obj/effect/ai_node/ai_node AS in GLOB.all_nodes)
+		nodes += list(ai_node.serialize())
+	rustg_register_nodes_astar(json_encode(nodes))
+
+#ifdef TESTING
+#define BENCHMARK_LOOP while(world.timeofday < end_time)
+#define BENCHMARK_RESET end_time = world.timeofday + duration
+
+/datum/controller/subsystem/advanced_pathfinding/proc/benchmark(run_number)
```
Comment (TiviPlus) [inline code/controllers/subsystem/advanced_pathfinding.dm:21]: Autodoc

File: code/controllers/subsystem/advanced_pathfinding.dm [hunk 8]
```diff
@@ -2,12 +2,50 @@ SUBSYSTEM_DEF(advanced_pathfinding)
...
+			dm_iterations++
+		BENCHMARK_RESET
+
+		BENCHMARK_LOOP
+			rustg_generate_path_astar("[start_node.unique_id]", "[goal_node.unique_id]")
+			rust_iterations++
+		BENCHMARK_RESET
+
+	message_admins("Average number of iterations for dm pathfinding in one sec : [dm_iterations/run_number]")
+	message_admins("Average number of iterations for rust pathfinding in one sec : [rust_iterations/run_number]")
```
Comment (TiviPlus) [inline code/controllers/subsystem/advanced_pathfinding.dm:46]: All the message admins

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Rest lgtm since rust is merged now

File: .github/workflows/ci_suite.yml (hunk 6 above)
Comment (BraveMole) [reply .github/workflows/ci_suite.yml:93]: No, I do that in the install_rust instead

**Pull request merged**